core/bin/external_node/src/config/mod.rs

```diff
@@ -66,7 +67,10 @@ impl RemoteENConfig {
                 .context("Failed to fetch L1 chain ID")?
                 .as_u64(),
         );
-
+        let base_token_addr = client
+            .get_base_token_l1_address()
+            .await
+            .context("Failed to fetch base token address")?;
```

Comment:
This probably wouldn't work if the new EN is released before the server version is updated (e.g. EN will attempt to call this method, but it doesn't exist yet on the older version of server) -- and this is likely to happen, because EN is almost always released first, since we have to test the new version on testnet for some time.
Probably what we want to do is to handle the error: if the request fails with "method does not exist" error, we fallback to the existing behavior (base token is ETH). We still can fail on other errors.

Reply:
Hi @popzxc!.

I've handled the case for the missing rpc method [here](https://github.com/matter-labs/zksync-era/pull/1031/files#diff-e52f3f81b2bcbc96468f7c93c1731a3a0c7aaab6a93c95b5e5e42173b1467740R74).
I'm not quite sure if this is the proper way to match with jsonrpsee's Error type, let me know if there's another and proper way to do it.

Also, one thing that to note is that I've hardcoded ETH's base token value [here](https://github.com/matter-labs/zksync-era/pull/1031/files#diff-e52f3f81b2bcbc96468f7c93c1731a3a0c7aaab6a93c95b5e5e42173b1467740R76), I'm guessing this should be properly set as a `const` value somewhere, but I haven't found where this should be done.

Reply:
Is `0x0000000000000000000000000000000000000001` an expected address?
We have L1 ETH address defined [here](https://github.com/matter-labs/zksync-era/blob/main/core/lib/constants/src/ethereum.rs#L6) and L2 ETH [here](https://github.com/matter-labs/zksync-era/blob/main/core/lib/constants/src/contracts.rs#L56) as constants, you should be able to reuse them.

Reply:
This is now set from the env, but the previous hardcoded value of`0x0000000000000000000000000000000000000001` was because the zk init set the env var CONTRACTS_BASE_TOKEN_ADDR=0x0000000000000000000000000000000000000001
when not specifiying a non-eth base token

Reply:
@fkrause98 Not sure if I understand. We want to fallback to ETH if there is no dedicated config set for that (this may happen when we'll be rolling out this version of server on existing zkSync Era deployments). Why would we use `0x0000000000000000000000000000000000000001 ` in that case? We want to use ETH address, no?

Reply:
The new changes for the shared bridge determine an eth-based chain based on the base token's address value.
For example, the sdk now has the following code:
- [isEthBasedChain function](https://github.com/zksync-sdk/zksync-ethers/blob/ethers-v5-feat/bridgehub/src/adapters.ts#L97).
- [ETH_ADDRESS_IN_CONTRACTS constant](https://github.com/zksync-sdk/zksync-ethers/blob/ethers-v5-feat/bridgehub/src/utils.ts#L34).

So, when a chain is eth-based, the base token address is 0x..01, else is a proper address to an erc20 contract on L1.

Reply:
But we're discussing the fallback for the case when shared bridge is not yet deployed, aren't we?
E.g. imagine that this feature is merged. This will happen, most likely, before we actually deploy a shared bridge and enable it for existing zkSync Era. The method will be there, but `0x00..01` does not represent an actual ETH token available back then.

I may be missing context, but right now it looks like we have 3 different cases:
1. Shared bridge is added, base token is ETH.
2. Shared bridge is added, base token is NOT ETH.
3. No shared bridge added (e.g. the development is still ongoing).

What I'm trying to explain is that we need to have sensible fallbacks for (3) if we want to merge it to main anytime soon.

Reply:
I think it makes sense to just make it return the zero address, it makes all 3 cases work and nothing really breaks from it so I would go that route. It is important to make sure users are aware of it though, as it means some APIs will return the zero address when the base token is `eth` and others (like the `baseToken` function in the new bridgehub) will return `0x00..01` instead.

Tagging @kelemeno for visibility as this pertains to contract/SDK changes.

Reply:
BTW, I've consulted @vladbochok on this topic and it appears that this is still WIP, but we indeed leaning towards having it as `0x00..01` everywhere, including public interfaces.

This unfortunately renders my proposal incorrect -- sorry for that.
Now I believe that the right thing would be to make the base token optional, e.g. the corresponding config variable would have an `Option<Address>` instead of `Address` and if it's `None` then in the API we'll return `Err(Web3Error::NotImplemented)`, otherwise we return whatever is stored in the config.

On EN we'll have two fallbacks then: first on `MethodNotFound` and second on `NotImplemented`, both resulting in setting base token to `None`. Otherwise it'll be `Some` whatever returned by API.

This way we'll have no additional hackery to support base token until it's explicitly enabled in existing envs.

Does it make sense?

Reply:
Yup, it does make sense, I'll make those changes and let you know.

Reply:
@popzxc I've implemented and tested the 3 cases you stated above and works as expected.

Reply:
@popzxc can we resolve this?